cpmpy/solvers/gurobi.py

```diff
@@ -0,0 +1,376 @@
+                # assumes that gurobi accepts sum(x) >= y without further simplification
+
+            # post the comparison
+            if cpm_expr.name == '<=':
+                return self.gbi_model.addLConstr(lhs, "<", rvar)
+            elif cpm_expr.name == '<':
+                raise Exception(f"{cpm_expr} should have been linearized, see /transformations/linearize.py")
+            elif cpm_expr.name == '>=':
+                return self.gbi_model.addLConstr(lhs, ">", rvar)
+            elif cpm_expr.name == '>':
+                raise Exception(f"{cpm_expr} should have been linearized, see /transformations/linearize.py")
+            elif cpm_expr.name == '!=':
+                raise Exception(f"{cpm_expr} should have been linearized, see /transformations/linearize.py")
+            elif cpm_expr.name == '==':
+                if not isinstance(lhs, Expression):
+                    # base cases: const|ivar|sum|wsum with prepped lhs above
+                    return self.gbi_model.addLConstr(lhs, "=", rvar)
+                elif lhs.name == "and":
+                    return self.gbi_model.addGenConstrAnd(rvar, self.solver_vars(lhs.args))
+                elif lhs.name == "or":
+                    return self.gbi_model.addGenConstrOr(rvar, self.solver_vars(lhs.args))
+                elif lhs.name == 'min':
+                    return self.gbi_model.addGenConstrMin(rvar, self.solver_vars(lhs.args))
+                elif lhs.name == 'max':
+                    return self.gbi_model.addGenConstrMax(rvar, self.solver_vars(lhs.args))
+                elif lhs.name == 'abs':
+                    # TODO put this in the correct place
+                    if isinstance(cpm_expr.args[1], _NumVarImpl):
+                        return self.gbi_model.addGenConstrAbs(rvar, self.solver_var(lhs.args[0]))
+                    # right side is a constant, not support by gurobi, so add new
+                    self += abs(lhs.args[0]) == intvar(rvar,rvar)
+                    return
+
+
+                elif lhs.name == 'mul':
+                    assert len(lhs.args) == 2, "Gurobi only supports multiplication with 2 variables"
+                    a, b = self.solver_vars(lhs.args)
+                    self.gbi_model.setParam("NonConvex",2)
+                    return self.gbi_model.addConstr(a * b == rvar)
+
+                elif lhs.name == 'div':
+                    if not isinstance(lhs.args[1], _NumVarImpl):
+                        a, b = self.solver_vars(lhs.args)
+                        return self.gbi_model.addLConstr(a / b, "=", rvar)
+                    raise Exception("Gurobi only supports division by constants")
+
+                elif lhs.name == 'pow':
+                    x, a = self.solver_vars(lhs.args)
+                    assert not isinstance(a, _NumVarImpl) or a.lb >= 0, f"Gurobi only supports power expressions with positive exponents."
+                    return self.gbi_model.addGenConstrPow(x, rvar, a)
+
```

Comment:
You are not using the single-reified wsum?
https://www.gurobi.com/documentation/9.5/refman/py_model_agc_indicator.html

that is a very powerful one that we will be able to use a lot I think... much better then bigM I would expect

Reply:
Yes, it seems that I have missed that one.
The implies constraint got translated automatically by get_or_make_var. I will fix it in the next commit

Reply:
On second tought, I don't think the indicator constraint of gurobi is directly applicable to our case. We cannot have something like (a -> b) == c as it gets flattened to (~a | b) == c

Reply:
I'm not sure you interpreted the docs right, it supports the following:
(BV == True) -> w * IVs >= 5
or
(BV = False) -> w * IVs = 1
etc

so BV -> lincons
for which I think you sometimes do a manual big-M in linearize.

Note that 'only_bv_implies' translates BV == lincons into two BV -> lincons
so the 'only_bv_implies' with a transformation that linearizes expressions to lincons and the above constraint is quite powerful for reified expressions...

Reply:
Ok, yes I indeed misinterpreted the docs.
We do lose the power of the "generalconstraints" if we apply the 'only_bv_implies' transformation tho.
At the moment we can post `a & b == c` to gurobi as `grb_model.addGenConstrAnd(c, [a,b])` But this `==` will of course be translated to implications. I am not sure which is the more optimal way to go here.

Reply:
Yes, we would lose that and don't want to lose that.

Perhaps we should add an 'eq_exept' argument that leaves constraints with that expr.name untouched in only_bv_implies 'BE == BV' part? This might be something we will need for other transformations too at some point, a list of 'only these' or 'all execpt these' names...

To be discussed. because only_bv_implies also transforms BoolExpr -> BV to ~BV -> ~BoolExpr, e.g. the Boolean literal at the left-hand side; I don't think you now support BoolExpr -> BV, so this part of only_bv_implies would also be usefl?